ros/angel_system_nodes/angel_system_nodes/audio_player.py

```diff
@@ -24,29 +23,54 @@ def __init__(self):
             1)
 
         self.audio_stream = bytearray()
+        self.audio_stream_lock = threading.RLock()
         self.audio_player_lock = threading.RLock()
         self.t = threading.Thread()
 
         self.audio_stream_duration = 0.0
         self.playback_duration = 1.0
 
+        self.prev_timestamp = None
+
     def listener_callback(self, msg):
         """
         Add the audio to the cumulative audio stream and spawns a new playback
         thread if the cumulative recording has reached more than a second and
         there is no ongoing playback.
         """
-        self.audio_stream.extend(msg.data)
-        self.audio_stream_duration += msg.sample_duration
+        log = self.get_logger()
+
+        # Check that this message comes temporally after the previous message
+        message_order_valid = False
+        if self.prev_timestamp is None:
+            message_order_valid = True
+        else:
+            if msg.header.stamp.sec > self.prev_timestamp.sec:
+                message_order_valid = True
+            elif msg.header.stamp.sec == self.prev_timestamp.sec:
+                # Seconds are same, so check nanoseconds
+                if msg.header.stamp.nanosec > self.prev_timestamp.nanosec:
+                    message_order_valid = True
+
+        if message_order_valid:
+            with self.audio_stream_lock:
+                self.audio_stream.extend(msg.data)
+                self.audio_stream_duration += msg.sample_duration
+        else:
+            log.info("Warning! Out of order messages.\n"
+                     + f"Prev: {self.prev_timestamp} \nCurr: {msg.header.stamp}")
+
+        self.prev_timestamp = msg.header.stamp
 
         if self.audio_stream_duration >= self.playback_duration and not(self.t.is_alive()):
             # Make a copy of the current data so we can play it back
             # while more data is accumulated
-            audio_data = copy.deepcopy(self.audio_stream)
+            audio_data = self.audio_stream
 
-            # Remove the data that we just copied from the stream
-            self.audio_stream = bytearray()
-            self.audio_stream_duration = 0.0
+            with self.audio_stream_lock:
```

Comment:
I think this lock with need to protect a larger scope. Currently, `audio_data = self.audio_stream` can be subject to clobbering by modifications currently occurring in critical sections. Also, `self.prev_timestamp = msg.header.stamp` is not protected, so what is actually assigned could be subject to race conditions. A naive thing to do is protect everything from order-check to `audio_data` assignment in the `self.audio_stream_lock` context, effectively serializing all calls to `listener_callback` (`log = ...` and thread spawning don't have to be in it I don't think).